### PR TITLE
Render variables starting with an underscore.

### PR DESCRIPTION
**Pull request by smoothml**
Currently, variables starting with an underscore (which are not shown to the user on project creation) are not rendered with Jinja2. If you want to create one of these variables based on other variables you cannot. For example, if you are creating a Python package repository and want to base the package name on the project name you might have a configuration like this

```python
{
    "project_name": "Project Name",
    "project_slug": "{{ cookiecutter.project_name|lower|replace(' ', '-') }}",
    "package_name": "{{ cookiecutter.project_name|lower|replace(' ', '_') }}",
}
```

Which could create a structure like this

```
project-name
├── Makefile
├── README.md
├── requirements.txt
└── src
    ├── project_name
    │   └── __init__.py
    ├── setup.py
    └── tests
        └── __init__.py
```

However, the user can change `project_slug` and `package_slug`. This may be desirable, but (as in my current situation) it may not be. You _could_ create post-generation hooks to fail project creation if these are changed, but this seems inelegant. This PR allows for the configuration to be specified like this

```python
{
    "project_name": "Project Name",
    "_project_slug": "{{ cookiecutter.project_name|lower|replace(' ', '-') }}",
    "_package_name": "{{ cookiecutter.project_name|lower|replace(' ', '_') }}",
}
```

The user then cannot change these variables, enforcing consistent naming.

### CODE REVIEW

**Comment by con-f-use** [comment]
I like this very much, as I ran into that as a relatively new user. Perhaps you should add a test for that as well.

**Comment by smoothml** [comment]
@con-f-use I'm glad you like it! You're right, I should add a test and some docs, though as far as I could see the underscore variables aren't explicitely tested for or documented (I didn't spend too long looking so may have missed it).

>have you checked how that interacts with the replay function / repo updater packages?

I haven't, but all the tests pass so I assume this is fine.

I'm not sure this will be merged given #1256 Perhaps this could go in the 2.0 release @insspb?

**Comment by insspb** [comment]
Hi @smoothml 
First of all - thank you for contribution. I really like this idea. Why I do not done it myself? Haha :)
Then here is my thoughts about it.

1. Remaking how current underscore works can break current templates. Maybe somebody read this code before and used it as feature for example with numbers. In your case numbers will be converted to string.
Here is code for you, to reproduce what I am talking about:
```python
    def test_should_render_dict(self):
        """

        """
        context = {
            'cookiecutter': {
                'project_name': 'Slartibartfast',
                '_project_name': 1123,
                'details': {'other_name': '{{cookiecutter.project_name}}'},
            }
        }

        cookiecutter_dict = prompt.prompt_for_config(context, no_input=True)
        assert cookiecutter_dict == {
            'project_name': 'Slartibartfast',
            '_project_name': 1123,
            'details': {'other_name': u'Slartibartfast',},
        }
```
This code will be good with current implementation. Will fail with your proposal. 

So...

Let do double underscore for user rendered private variables + docs + test. And it can be included in nearest release after this will be done.

But I really like this idea. At least things like this will be shorter: https://github.com/ionelmc/cookiecutter-pylibrary/blob/master/cookiecutter.json

Currently included in 1.7.1, but as we are plan to make a release in few days maybe it will not be ready for that time. At this case i will move it to 2.0.

**Comment by smoothml** [comment]
Thanks for your comments @insspb. Good point regarding numbers vs. strings. Double underscores work 👍 I'm not sure I'll be able to get this ready in the next few days so getting it in 2.0 is a good aim I think.

**Comment by ssbarnea** [comment]
I like this change but please add some tests for it and we should merge it in next major release 1.8.0, not the patch one (1.7.x) due to the risk of breaking existing templates.

The risk of breaking templates is small and template authors should already have CI/CD configured to test their templates with cookiecutter from master, running jobs at least weekly. We will always wait at least one week before merging a breaking change and tagging a release or pre-release.

File: tests/test_prompt.py [hunk 1]
```diff
@@ -260,6 +260,29 @@ def test_dont_prompt_for_private_context_var(self, monkeypatch):
         cookiecutter_dict = prompt.prompt_for_config(context)
         assert cookiecutter_dict == {'_copy_without_render': ['*.html']}
 
+    def test_render_hidden_variables(self):
```
Comment (insspb) [inline tests/test_prompt.py:263]: Docstrings missed.

Comment (smoothml) [reply tests/test_prompt.py:263]: Yup, thanks. I haven't finished yet!

Comment (smoothml) [reply tests/test_prompt.py:263]: I should have made it a draft PR actually. I'll do that.

**Comment by insspb** [comment]
@smoothml still draft?

**Comment by smoothml** [comment]
Yeah, sorry. Been slammed the last couple of weeks. Just docs to go. Will try and get it done this weekend. Sorry for the delay!

**Comment by smoothml** [comment]
@insspb @ssbarnea apologies for the delay. This is ready for review now.

**Comment by con-f-use** [comment]
We seem to have a [regression here](https://github.com/cookiecutter/cookiecutter/issues/1831). This happens on the latest 2.1.1 version. @ssbarnea

**Pull request merged**